### PR TITLE
Tracy Profiling

### PR DESCRIPTION
**Pull request by zach2good**
This is the power of CMake!

Build with `TRACY_ENABLE` to turn on the profiling, otherwise, everything will get stripped away and has no effect on the build.

This PR adds support for the Tracy client in topaz_game. Once the game is built and running, you can connect the server to it to see the beautiful stats in realtime:

image

We are hardcoded to v0.7.3, different versions of client and server are not compatible!

https://github.com/wolfpld/tracy/releases/tag/v0.7.3

**GUIDE**

https://github.com/project-topaz/topaz/wiki/Performance-Profiling-with-Tracy

**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: src/map/entities/charentity.cpp [hunk 1]
```diff
@@ -255,6 +255,12 @@ void CCharEntity::clearPacketList()
 
 void CCharEntity::pushPacket(CBasicPacket* packet)
 {
+    ZoneScoped;
+    const char* cstr = reinterpret_cast<const char*>(GetName());
+    ZoneText(cstr, std::strlen(cstr));
+    auto packet_id = fmt::format("{:2x}", packet->id());
+    ZoneText(packet_id.c_str(), packet_id.size());
```
Comment (zach2good) [inline src/map/entities/charentity.cpp]: Note to self:
```cpp
std::string Hex8ToString(uint8 hex)
{
    return fmt::format("0x{:2X}", hex);
}

std::string Hex16ToString(uint16 hex)
{
    return fmt::format("0x{:4X}", hex);
}
```

Comment (zach2good) [reply src/map/entities/charentity.cpp]:
```cpp
#define ZoneString(str) ZoneText(str.c_str(), str.size())
#define ZoneCString(cstr) ZoneText(cstr, std::strlen(cstr))
#define ZoneIString(istr) ZoneCString(reinterpret_cast<const char*>(istr))
```

**Comment by zach2good** [comment]
Note to self:
```
To report data, use the TracyPlot(name, value) macro.
To configure how plot values are presented by the profiler, you may use the TracyPlotConfig(name,
format) macro, where format is one of the following options:
• tracy::PlotFormatType::Number – values will be displayed as plain numbers.
• tracy::PlotFormatType::Memory – treats the values as memory sizes. Will display kilobytes,
megabytes, etc
```

```cpp
//what:
//LUA_GCCOUNT: returns the current amount of memory (in Kbytes) in use by Lua.
//LUA_GCCOUNTB: returns the remainder of dividing the current amount of bytes of memory in use by Lua by 1024.
//LUA_API int (lua_gc) (lua_State *L, int what, int data);

#ifdef TRACY_ENABLE
TracyPlotConfig("Lua Memory Usage", tracy::PlotFormatType::Memory);
TracyPlot("Lua Memory Usage", lua_gc(L, LUA_GCCOUNT, 0) / 1024); // from kb to b
#endif
```

**Review by ibm2431 — APPROVED** [review]
Let us embark on a new era of performance testing!

**Pull request merged**